boxtree/tree_build.py

```diff
@@ -46,6 +58,273 @@ class MaxLevelsExceeded(RuntimeError):
+    applied to the last dimension.
+    """
+    old_size = arr.shape[-1]
+    prefix = (slice(None), ) * (arr.ndim - 1)
+    if old_size >= new_size:
+        return arr[prefix + (slice(new_size), )].copy()
+    else:
+        new_shape = list(arr.shape)
+        new_shape[-1] = new_size
+        new_arr = np.zeros(new_shape, arr.dtype)
+        new_arr[prefix + (slice(old_size), )] = arr
+        return new_arr
+
+
+def _vec_of_signs(dim: int, i: int) -> npt.NDArray:
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
+    n = len(binary_digits)
+    assert n <= dim
+    return np.array([0]*(dim-n) + binary_digits) * 2 - 1
+
+
+# {{{ refine/coarsen a tree of boxes
+
+def refined(tob: TreeOfBoxes, refine_flags: npt.NDArray) -> TreeOfBoxes:
+    """Make a refined copy of `tob` where boxes flagged with `refine_flags` are
+    refined.
+    """
+    return refined_and_coarsened(tob, refine_flags, None)
+
+
+def uniformly_refined(tob: TreeOfBoxes) -> TreeOfBoxes:
+    """Make a uniformly refined copy of `tob`.
+    """
+    refine_flags = np.zeros(tob.nboxes, bool)
+    refine_flags[tob.get_leaf_flags()] = 1
+    return refined(tob, refine_flags)
+
+
+def coarsened(tob: TreeOfBoxes, coarsen_flags: npt.NDArray,
+              error_on_ignored_flags: bool = True
+              ) -> TreeOfBoxes:
+    """Make a coarsened copy of `tob` where boxes flagged with `coarsen_flags`
+    are coarsened.
+    """
+    return refined_and_coarsened(
+        tob, None, coarsen_flags, error_on_ignored_flags)
+
+
+def _apply_refine_flags_without_sorting(refine_flags, tob):
+    if refine_flags[~tob.get_leaf_flags()].any():
+        raise ValueError("attempting to split non-leaf")
+
+    refine_parents, = np.where(refine_flags)
+    if len(refine_parents) == 0:
+        return tob
+
+    nchildren = 2**tob.dim
+    n_new_boxes = len(refine_parents) * nchildren
+    nboxes_new = tob.nboxes + n_new_boxes
+
+    child_box_starts = (
+            tob.nboxes
+            + nchildren * np.arange(len(refine_parents)))
+
+    refine_parents_per_child = np.empty(
+            (nchildren, len(refine_parents)), np.intp)
+    refine_parents_per_child[:] = refine_parents.reshape(-1)
+    refine_parents_per_child = refine_parents_per_child.reshape(-1)
+
+    box_parents = _resized_array(tob.box_parent_ids, nboxes_new)
+    box_centers = _resized_array(tob.box_centers, nboxes_new)
+    box_children = _resized_array(tob.box_child_ids, nboxes_new)
+    box_levels = _resized_array(tob.box_levels, nboxes_new)
+
+    # new boxes are appended at the end, so applying coarsen_flags wrt the
+    # original tree is still meaningful after this
+    box_parents[tob.nboxes:] = refine_parents_per_child
+    box_levels[tob.nboxes:] = tob.box_levels[box_parents[tob.nboxes:]] + 1
+    box_children[:, refine_parents] = (
+        child_box_starts + np.arange(nchildren).reshape(-1, 1))
```

Comment:
Here's a potentially bigger concern: The traversal and various parts of the data structure assume that levels are contiguous in box number (as evidenced by things like `level_start*box_nrs`. How does this code deal with that, if at all?

Reply:
In `refined_and_coarsened`, `_sort_and_prune_deleted_boxes` is called after coarsening, which sorts boxes by levels.

Reply:
Resolved: None of those are exposed to the user, so :shrug:, let's not.